synapse/handlers/room_member.py

```diff
@@ -218,6 +220,52 @@ def _local_membership_update(
                 newly_joined = prev_member_event.membership != Membership.JOIN
             if newly_joined:
                 yield self._user_joined_room(target, room_id)
+
+            # Copy over direct message status and room tags if this is a join
```

Comment:
please can this stuff move out to a separate function? as a rule of thumb, try to keep functions within a screenful.